Add new “Shallow copy” and “Deep copy” glossary entries (#12474)
mdn/content: files/en-us/glossary/deep_copy/index.md

---
title: Deep copy
slug: Glossary/Deep_copy
tags:
  - Glossary
  - Deep copy
---
{{MDNSidebar}}

A **deep copy** of an object is a copy whose members do not share the same references as those of the object from which the copy was made — in contrast with a [shallow copy](/en-US/docs/Glossary/Shallow_copy). Standard built-in object-copy operations in JavaScript do not create deep copies (instead, they create shallow copies).

> When you want a deep copy of a JavaScript object, you can make one by using {{jsxref("JSON.stringify()")}} to convert the object to a JSON string, and then {{jsxref("JSON.parse()")}} to convert the string to a JavaScript object:

Reviewer: obj->JSON->obj is a way to make a deep copy, but there are many, many cases where it will fail. Functions (with closures), Symbols, HTML elements, recursive data, ...

I would refer to the limitations of JSON.stringify / parse.

This article is obviously difficult to write simply and exhaustively.

Author: > obj->JSON->obj is a way to make a deep copy, but there are many, many cases where it will fail.

Important point, yeah — thanks, it definitely should be mentioned. So I pushed an update that includes language that I hope makes it clear (without attempting to enumerate all the reasons that might cause an object to be unserializable).

Author: > Important point, yeah — thanks, it definitely should be mentioned. So I pushed an update that includes language that I hope makes it clear (without attempting to enumerate all the reasons that might cause an object to be unserializable).

Actually, in a subsequent change, based on additional review comments, I did end up adding a list of some cases where objects aren’t serializable (based on the list in https://github.com/mdn/content/pull/12474#discussion_r794909367). If it would be useful to add more to that list, I guess we can do that.

---
title: Deep copy
slug: Glossary/Deep_copy
tags:
  - Glossary
  - Deep copy
---
{{MDNSidebar}}

> 

Reviewer 2: If you go with my suggestion in https://github.com/mdn/content/pull/12474/files#r796173795 you might want to make similar changes here:
1. Make it clear that "not share the same references" means "completely independent object"
2. The risk of shallow copies is unintentional modification of the original. That's worth having up near the top.
3. Moving "In JavaScript, standard built-in .." to the end of intro makes sense because you don't care about this until you really understand what a deep copy is.

Author: > If you go with my suggestion in [#12474 (files)](https://github.com/mdn/content/pull/12474/files#r796173795) you might want to make similar changes here:
> 
>     1. Make it clear that "not share the same references" means "completely independent object"
> 
>     2. The risk of shallow copies is unintentional modification of the original. That's worth having up near the top.
> 
>     3. Moving "In JavaScript, standard built-in .." to the end of intro makes sense because you don't care about this until you really understand what a deep copy is.

OK, pushed an update that includes those changes

A **deep copy** of an object is a copy whose properties do not share the same references as those of the source object from which the copy was made — in contrast with a [shallow copy](/en-US/docs/Glossary/Shallow_copy). In JavaScript, standard built-in object-copy operations ([`Object.assign()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign), [`Array.from()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from), [`Array.prototype.slice()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice), and [spread syntax](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)) do not create deep copies (instead, they create shallow copies).

> To make a deep copy of a JavaScript object, you may be able to use {{jsxref("JSON.stringify()")}} to serialize the object to a JSON string, and then {{jsxref("JSON.parse()")}} to convert the string to a JavaScript object:

Reviewer 2: This is good, but at this point if you keep the text above I might not really understand what a deep copy really means and why serializing the object will help. When can't I serialize and object? Should you use stringify or structuredClone?

Author: > This is good, but at this point if you keep the text above I might not really understand what a deep copy really means and why serializing the object will help. When can't I serialize and object?

Added some examples of cases where objects can’t be serialized.

> Should you use stringify or structuredClone?

Added mention that `structuredClone()` allows transferable objects to be transferred rather than being cloned. I guess that’s the main selling point that of `structuredClone()` over stringifying and reparsing.

```js
let ingredients_list = ["noodles",{"list":["eggs","flour","water"]}];
let ingredients_list_deepcopy = JSON.parse(JSON.stringify(ingredients_list));

// Change the value of the 'list' property in ingredients_list_copy.
ingredients_list_deepcopy[1].list = ["rice flour","water"]
// The 'list' property does not change in ingredients_list.
console.log(ingredients_list[1].list);
// Array(3) [ "eggs", "flour", "water" ]
```

As can be seen from the code above, because a deep copy shares no references with its source object, any changes made to the deep copy do not affect the source object.

> However, while it’s easy to see that an object such as the one in the example above is serializable, many JavaScript objects are not serializable at all — and so, calling `JSON.stringify()` to serialize them will fail. So there’s no way to make deep copies of such objects.

Reviewer 3: I'm not very comfortable telling folks that their subjective experience of these things is necessarily easy. I had to read this twice to be sure I understood what you were saying, at least. And removing the aside about ease seems to tighten this up a fair amount. What do you think?

```suggestion
However, many JavaScript objects are not serializable at all and calling `JSON.stringify()` to serialize them will fail. So there’s no way to make deep copies of such objects.
```

> For objects that _are_ serializable, you can alternatively use the [`window.structuredClone()`](/en-US/docs/Web/API/structuredClone) method to create deep copies. But note that `window.structuredClone()` isn’t a feature of the JavaScript language itself — instead it’s a feature of browsers and any other JavaScript runtimes that implement the global [`window`](/en-US/docs/Web/API/Window) object. And calling `window.structuredClone()` to clone a non-serializable object will fail in the same way that calling `JSON.stringify()` to serialize it will fail.

Reviewer 2: structuredClone can be used in workers too.

```suggestion
For objects that _are_ serializable, you can alternatively use the [`structuredClone()`](/en-US/docs/Web/API/structuredClone) method to create deep copies. But note that `structuredClone()` isn’t a feature of the JavaScript language itself — instead it’s a feature of browsers and any other JavaScript runtimes that implement a global object like [`window`](/en-US/docs/Web/API/Window). And calling `structuredClone()` to clone a non-serializable object will fail in the same way that calling `JSON.stringify()` to serialize it will fail.
```

---
title: Deep copy
slug: Glossary/Deep_copy
tags:
  - Glossary
  - Deep copy
---
{{MDNSidebar}}

> A **deep copy** of an object is a copy whose properties do not share the same references (point to the same underlying values) as those of the source object from which the copy was made. As a result, when you change either the source or the copy, you can be assured you’re not causing the other object to change too; that is, you won’t unintentionally be causing changes to the source or copy that you don’t expect. That behavior contrasts with the behavior of a [shallow copy](/en-US/docs/Glossary/Shallow_copy), in which changes to either the source or the copy may also cause the other object to change too (because the two objects share the same references).

Reviewer 2: This is all much better than before in both deep and shallow copy. If it were me I might write "(point to the same underlying data)" rather than values - mostly because I think of a value as a number, and here you are using it like "memory location". Ditto shallow copy.

In JavaScript, standard built-in object-copy operations ([spread syntax](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), [`Array.prototype.slice()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice), [`Array.from()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from), [`Object.assign()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)) do not create deep copies (instead, they create shallow copies).

> To make a deep copy of a JavaScript object, you may be able to use {{jsxref("JSON.stringify()")}} to [serialize](/en-US/docs/Glossary/Serialization) the object to a JSON string, and then {{jsxref("JSON.parse()")}} to convert the string to a JavaScript object:

Reviewer 2: Mabye
```suggestion
One way to make a deep copy of a JavaScript object, if it can be [serialized](/en-US/docs/Glossary/Serialization), is to use {{jsxref("JSON.stringify()")}} to convert the object to a JSON string, and then {{jsxref("JSON.parse()")}} to convert the string back into (a completely new) JavaScript object:
```

> One way to make a deep copy of a JavaScript object, if it can be [serialized](/en-US/docs/Glossary/Serialization), is to use {{jsxref("JSON.stringify()")}} to convert the object to a JSON string, and then {{jsxref("JSON.parse()")}} to convert the string back into (a completely new) JavaScript object:

Reviewer 2: Sorry, I stuffed that.
```suggestion
One way to make a deep copy of a JavaScript object, if it can be [serialized](/en-US/docs/Glossary/Serialization), is to use {{jsxref("JSON.stringify()")}} to convert the object to a JSON string, and then {{jsxref("JSON.parse()")}} to convert the string back into a (completely new) JavaScript object:
```
